README.md
# mb.scrollable 1.5.7

__An open source jQuery component to easily build a scrollable group of elements.__

![mb.extruder](http://pupunzi.com/gitHub/mb.scrollable.jpg)

## [go to the demo](http://pupunzi.com/#mb.components/mb.scrollable/scrollable.html)
## [go to the doc](http://wiki.github.com/pupunzi/jquery.mb.scrollable/)
## [go to the blog](http://pupunzi.open-lab.com/mb-jquery-components/jquery-mb-scrollable/)


[jquery.mb.components](http://pupunzi.com/), another way of thinking the web


inc/changeLog.txt


jQuery.mb.scrollable
©2002 - 2009 Matteo Bicocchi

1.7.0 reformat code:
  Cleaned up the methods to get the index (now all refers to the gotoPage method);
  added “loadCallback”, “changePageCallback”.

1.6.0 Safari bugfix:
  adjusted dimensions to prevent non int value.

1.5.7 minor ISSUE:
  added "goToPage" method. Allow to jump to a specific page.


1.5.5 minor ISSUE:
  added two callback functions: nextCallback and prevCallback.
  solved an align problem. 

1.5.2 minor fix for IE7:
  added float left to elements in the component for the unsupported inline-block attr on display.

1.5.1 minor fix:
  changed "dir:orizontal" in "dir:horizontal".

1.5 Major issue:
  introduced the vertical scroll modality controlled by a new parameter "dir".

1.0 First release

inc/mbScrollable.pack.js
/*******************************************************************************
 jquery.mb.components
 Copyright (c) 2001-2010. Matteo Bicocchi (Pupunzi); Open lab srl, Firenze - Italy
 email: info@pupunzi.com
 site: http://pupunzi.com

 Licences: MIT, GPL
 http://www.opensource.org/licenses/mit-license.php
 http://www.gnu.org/licenses/gpl.html
 ******************************************************************************/

/*
 * Name:jquery.mb.scrollable
 * Version: 1.5.7
*/
eval(function(p,a,c,k,e,r){e=function(c){return(c<a?'':e(parseInt(c/a)))+((c=c%a)>35?String.fromCharCode(c+29):c.toString(36))};if(!''.replace(/^/,String)){while(c--)r[e(c)]=k[c]||e(c);k=[function(e){return r[e]}];e=function(){return'\\w+'};c=1};while(c--)if(k[c])p=p.replace(new RegExp('\\b'+e(c)+'\\b','g'),k[c]);return p}('(8($){$.q={1D:"1E.1F",1G:"1H",1I:"1.0",1l:{1m:"1n",j:1J,k:4,t:2,N:B,r:"14",O:"#O",u:1K,P:B,1o:1L,15:8(){},16:8(){}},1p:8(f){Y h.1q(8(){h.3={};$.1r(h.3,$.q.1l);$.1r(h.3,f);9 a=h;a.D=a.3.1m!="1n";9 b=$(a.3.O);a.l=1;a.6=0;a.m=$(a).17();a.m.G("1s");b.1t();$(a).17().1q(8(){$(h).1M("<18 19=\'1N\'></18>")});7(a.3.N){$(a.m).I("-1O-1u-N",a.3.N);$(a.m).I("-1P-1u-N",a.3.N)}a.m=$(a).17();9 c=a.3.k<h.m.J()?a.3.k:a.m.J();7(a.D){a.E=(a.3.r/c)-a.3.t;$(a.m).I({1Q:a.3.t,r:a.E,j:a.3.j})}v{a.E=(a.3.j/c)-a.3.t;$(a.m).I({1R:a.3.t,j:a.E,1S:"1T-1U",1V:"1W"})}h.o=(a.E*a.m.J())+(a.3.t*a.m.J());a.F=1a.1X(a.m.J()/a.3.k);9 d=$.1b.1c&&a.3.k>2?a.3.t/(a.3.k):0;7(a.D)$(a).I({1v:"1w",r:a.3.r-d,j:a.3.j,1Y:5,1d:"1e"});v $(a).I({1v:"1w",j:a.3.j-d,r:a.3.r,1Z:5,1d:"1e"});9 e=$("<18 19=\'1x\'/>").I({j:a.o,1d:"1e"});$(a.m).20(e);a.w=$(a).g(".1x");$(a.m).21(8(){7(a.z)$(a).A()},8(){7(a.z)$(a).Q()});7(a.3.P&&a.m.J()>a.3.k){a.z=V;$(a).Q()}$(a).Z();$(a).10();22(8(){$(".1s").1y()},23);$(a).s()})},W:8(a){9 b=$(h).C(0);7(!a)b.z=B;7(b.l==b.F){$(b).s();Y}7(b.3.15)b.3.15();9 c=$.1b.1c&&b.3.k>2?b.3.t/b.3.k:0;b.6-=((b.E+b.3.t)*b.3.k)-c;7(b.D){7((b.6<-b.o+b.3.r))b.6=-b.o+b.3.r;$(b.w).x({R:b.6},b.3.u)}v{7((b.6<-b.o+b.3.j))b.6=-b.o+b.3.j;$(b.w).x({S:b.6},b.3.u)}b.l+=1;$(h).s()},X:8(a){9 b=$(h).C(0);7(!a)b.z=B;7(b.l==1){$(b).s();Y}7(b.3.16)b.3.16();9 c=$.1b.1c&&b.3.k>2?b.3.t/b.3.k:0;b.6+=((b.E+b.3.t)*b.3.k)+c;7(b.D){7(b.6>=0)b.6=0;$(b.w).x({R:b.6},b.3.u)}v{7(b.6>=0)b.6=0;$(b.w).x({S:b.6},b.3.u)}b.l-=1;$(h).s()},11:8(){9 a=$(h).C(0);a.z=B;a.6=0;7(a.D){$(a.w).x({R:a.6},a.3.u)}v{$(a.w).x({S:a.6},a.3.u)}a.l=1;$(h).s();$(a).A()},12:8(){9 a=$(h).C(0);a.z=B;7(a.D){a.6=-a.o+a.3.r;$(a.w).x({R:a.6},a.3.u)}v{a.6=-a.o+a.3.j;$(a.w).x({S:a.6},a.3.u)}a.l=a.F;$(a).s();$(a).A()},Z:8(){9 a=$(h).C(0);9 b=$(a.3.O);9 c=b.g(".24");7(c){8 1z(i){a.6=-((a.E+a.3.t)*(a.3.k*(i-1)));7(a.D){7(a.6<-a.o+a.3.r)a.6=-a.o+a.3.r;$(a.w).x({R:a.6},a.3.u)}v{7(a.6<-a.o+a.3.j)a.6=-a.o+a.3.j;$(a.w).x({S:a.6},a.3.u)}a.l=1a.1A(i);$(a).s()}9 n=0;25(9 i=1;i<=a.F;i++){n++;9 p=$("<1B 19=\'1f\'> "+n+" <\\/1B>").K("L",8(){1z($(h).1C());$(a).A();a.z=B});c.26(p)}}},Q:8(){9 a="T";9 b=$(h).C(0);b.z=V;7(b.P)Y;9 c=b.3.1o+b.3.u;b.P=V;b.14=27(8(){a=b.l==1?"T":b.l==b.F?"13":a;7(a=="T")$(b).W(V);v $(b).X(V)},c);$(b).s()},A:8(){9 a=$(h).C(0);a.P=B;28(a.14);$(a).s()},10:8(){9 a=$(h).C(0);9 b=$(a.3.O);b.g(".1g").K("L",8(){$(a).11()});b.g(".13").K("L",8(){$(a).A();$(a).X()});b.g(".T").K("L",8(){$(a).A();$(a).W()});b.g(".1h").K("L",8(){$(a).12()});b.g(".1i").K("L",8(){$(a).Q()});b.g(".1j").K("L",8(){$(a).A();a.z=B})},s:8(){9 a=$(h).C(0);9 b=$(a.3.O);7(a.m.J()<=a.3.k){b.1t()}v{b.1y()}7(a.l==a.F){b.g(".1h").G("H");b.g(".T").G("H")}v{b.g(".1h").M("H");b.g(".T").M("H")}7(a.l==1){b.g(".1g").G("H");b.g(".13").G("H")}v{b.g(".1g").M("H");b.g(".13").M("H")}7(a.P){b.g(".1i").G("U");b.g(".1j").M("U")}v{b.g(".1i").M("U");b.g(".1j").G("U")}b.g(".1f").M("U");b.g(".1f").29(a.l-1).G("U");b.g(".l").1C(a.l+" / "+a.F)},1k:8(i,a){9 b=$(h).C(0);9 c=a?0:b.3.u;7(i>b.F)i=b.F;b.6=-((b.E+b.3.t)*(b.3.k*(i-1)));7(b.D){7(b.6<-b.o+b.3.r)b.6=-b.o+b.3.r;$(b.w).x({R:b.6},c)}v{7(b.6<-b.o+b.3.j)b.6=-b.o+b.3.j;$(b.w).x({S:b.6},c)}b.l=1a.1A(i);$(b).s();$(b).A();b.z=B}};$.y.q=$.q.1p;$.y.W=$.q.W;$.y.X=$.q.X;$.y.11=$.q.11;$.y.12=$.q.12;$.y.Z=$.q.Z;$.y.2a=$.q.2b;$.y.Q=$.q.Q;$.y.A=$.q.A;$.y.10=$.q.10;$.y.s=$.q.s;$.y.1k=$.q.1k})(2c);',62,137,'|||options|||scrollTo|if|function|var|||||||find|this||width|elementsInPage|idx|elements||elementsDim||mbScrollable|height|mbManageControls|elementMargin|slideTimer|else|mbscrollableStrip|animate|fn|autoScrollActive|mbStopAutoscroll|false|get|isVertical|singleElDim|totalPages|addClass|disabled|css|size|bind|click|removeClass|shadow|controls|autoscroll|mbAutoscroll|marginTop|marginLeft|next|sel|true|mbNextPage|mbPrevPage|return|mbPageIndex|mbActivateControls|mbFirstPage|mbLastPage|prev|auto|nextCallback|prevCallback|children|div|class|Math|browser|safari|position|relative|page|first|last|start|stop|goToPage|defaults|dir|horizontal|scrollTimer|buildMbScrollable|each|extend|scrollEl|hide|box|overflow|hidden|scrollableStrip|fadeIn|getPage|floor|span|html|plugin|mb|scroller|author|MB|version|950|600|6000|wrap|SECont|moz|webkit|marginBottom|marginRight|display|inline|block|float|left|ceil|paddingRight|paddingBottom|wrapAll|hover|setTimeout|1000|pageIndex|for|append|setInterval|clearInterval|eq|mbgotoPage|gotoPage|jQuery'.split('|'),0,{}))

inc/mbScrollable.js
/*******************************************************************************
 jquery.mb.components
 Copyright (c) 2001-2010. Matteo Bicocchi (Pupunzi); Open lab srl, Firenze - Italy
 email: mbicocchi@open-lab.com
 site: http://pupunzi.com

 Licences: MIT, GPL
 http://www.opensource.org/licenses/mit-license.php
 http://www.gnu.org/licenses/gpl.html
 ******************************************************************************/

/*
 * Name:jquery.mb.scrollable
 * Version: 1.7.0
 */

(function($) {
	$.mbScrollable= {
		plugin:"mb.scrollable",
		author:"Matteo Bicocchi",
		version:"1.7.0",
		defaults:{
			dir:"horizontal",
			textDir:"ltr",
			width:980,
			elementsInPage:4,
			elementMargin:2,
			shadow:false,
			height:"auto",
			controls:"#controls",
			slideTimer:600,
			autoscroll:false,
			scrollTimer:6000,

			loadCallback:function(){},
			nextCallback:function(){},
			prevCallback:function(){},
			changePageCallback:function(){}
		},

		buildMbScrollable: function(options){
			return this.each (function (){
				this.options = {};
				$.extend (this.options, $.mbScrollable.defaults);
				$.extend (this.options, options);

				var mbScrollable=this;
				mbScrollable.isVertical= mbScrollable.options.dir!="horizontal";
				var controls=$(mbScrollable.options.controls);
				mbScrollable.idx=1;
				mbScrollable.scrollTo=0;
				mbScrollable.elements= $(mbScrollable).children();
				mbScrollable.elements.addClass("scrollEl");
				controls.hide();

				$(mbScrollable).children().each(function(){$(this).wrap("<div class='SECont'></div>");});
				if (mbScrollable.options.shadow){
					$(mbScrollable.elements).css("-moz-box-shadow",mbScrollable.options.shadow);
					$(mbScrollable.elements).css("-webkit-box-shadow",mbScrollable.options.shadow);
				}
				mbScrollable.elements= $(mbScrollable).children();
				var eip= mbScrollable.options.elementsInPage<this.elements.size()?mbScrollable.options.elementsInPage:mbScrollable.elements.size();
				if(mbScrollable.isVertical){
					mbScrollable.singleElDim= Math.floor((mbScrollable.options.height/eip)-mbScrollable.options.elementMargin);
					$(mbScrollable.elements).css({marginBottom:mbScrollable.options.elementMargin, height:mbScrollable.singleElDim, width:mbScrollable.options.width});
				}else{
					mbScrollable.singleElDim= Math.floor((mbScrollable.options.width/eip)-mbScrollable.options.elementMargin);
					$(mbScrollable.elements).css({marginRight:mbScrollable.options.elementMargin, width:mbScrollable.singleElDim, display:"inline-block","float":"left" });
				}
				this.elementsDim= (mbScrollable.singleElDim*mbScrollable.elements.size())+(mbScrollable.options.elementMargin*mbScrollable.elements.size());
				mbScrollable.totalPages= Math.ceil(mbScrollable.elements.size()/mbScrollable.options.elementsInPage);

				if(mbScrollable.isVertical)
					$(mbScrollable).css({overflow:"hidden", height:((mbScrollable.singleElDim+mbScrollable.options.elementMargin)*mbScrollable.options.elementsInPage), paddingRight:5, position:"relative"});
				else
					$(mbScrollable).css({overflow:"hidden", width:((mbScrollable.singleElDim+mbScrollable.options.elementMargin)*mbScrollable.options.elementsInPage),height:mbScrollable.options.height,paddingBottom:5, position:"relative"});

				var mbScrollableStrip=$("<div class='scrollableStrip'/>").css({width:mbScrollable.elementsDim, position:"relative"});
				$(mbScrollable.elements).wrapAll(mbScrollableStrip);
				mbScrollable.mbscrollableStrip=$(mbScrollable).find(".scrollableStrip");
				$(mbScrollable.elements).hover(
						function(){
							if(mbScrollable.autoScrollActive)
								$(mbScrollable).mbStopAutoscroll();
						},
						function(){
							if(mbScrollable.autoScrollActive)
								$(mbScrollable).mbAutoscroll();
						});
				if(mbScrollable.options.autoscroll && mbScrollable.elements.size()>mbScrollable.options.elementsInPage){
					mbScrollable.autoScrollActive=true;
					$(mbScrollable).mbAutoscroll();
				}
				$(mbScrollable).mbPageIndex();
				$(mbScrollable).mbActivateControls();
				setTimeout(function(){
					$(".scrollEl").fadeIn();
				},1000);
				$(mbScrollable).mbManageControls();
			});
		},

		mbNextPage: function(auto){
			var mbScrollable= $(this).get(0);
			if (!auto) mbScrollable.autoScrollActive=false;

			if(mbScrollable.idx==mbScrollable.totalPages){
				$(mbScrollable).mbManageControls();
				return;
			}
			mbScrollable.idx+=1;
			$(mbScrollable).goToPage(mbScrollable.idx,false);
			if(mbScrollable.options.nextCallback)
				mbScrollable.options.nextCallback(mbScrollable);
		},

		mbPrevPage: function(auto){
			var mbScrollable= $(this).get(0);
			if (!auto) mbScrollable.autoScrollActive=false;

			if(mbScrollable.idx==1){
				$(mbScrollable).mbManageControls();
				return;
			}

			mbScrollable.idx-=1;
			$(mbScrollable).goToPage(mbScrollable.idx,false);
			if(mbScrollable.options.prevCallback)
				mbScrollable.options.prevCallback(mbScrollable);

		},

		mbFirstPage: function(){
			var mbScrollable= $(this).get(0);

			mbScrollable.idx=1;
			$(mbScrollable).goToPage(mbScrollable.idx,false);

		},

		mbLastPage: function(){
			var mbScrollable= $(this).get(0);
			mbScrollable.idx=mbScrollable.totalPages;
			$(mbScrollable).goToPage(mbScrollable.idx,false);
		},

		mbPageIndex: function(){
			var mbScrollable= $(this).get(0);
			var controls=$(mbScrollable.options.controls);
			var pages=controls.find(".pageIndex");
			if (pages){
				var n=0;
				for(var i=1;i<=mbScrollable.totalPages;i++){
					n++;
					var p=$("<span class='page'> "+n+" <\/span>").bind("click",function(){
						mbScrollable.autoScrollActive=false;
						$(mbScrollable).goToPage($(this).html(),false)
					});
					pages.append(p);
				}
			}
		},
		mbAutoscroll:function(){
			var dir= "next";
			var mbScrollable= $(this).get(0);
			mbScrollable.autoScrollActive=true;

			if(mbScrollable.autoscroll) return;
			var timer=mbScrollable.options.scrollTimer+mbScrollable.options.slideTimer;
			mbScrollable.autoscroll = true;
			mbScrollable.auto = setInterval(function(){
				dir= mbScrollable.idx==1?"next":mbScrollable.idx==mbScrollable.totalPages?"prev":dir;
				if(dir=="next")
					$(mbScrollable).mbNextPage(true);
				else
					$(mbScrollable).mbPrevPage(true);
			},timer);
			$(mbScrollable).mbManageControls();
		},

		mbStopAutoscroll: function(){
			var mbScrollable= $(this).get(0);
			mbScrollable.autoscroll = false;
			clearInterval(mbScrollable.auto);
			$(mbScrollable).mbManageControls();

		},

		mbActivateControls: function(){
			var mbScrollable=$(this).get(0);

			if(mbScrollable.options.loadCallback)
				mbScrollable.options.loadCallback(mbScrollable);

			var controls=$(mbScrollable.options.controls);
			controls.find(".first").bind("click",function(){$(mbScrollable).mbFirstPage();});
			controls.find(".prev").bind("click",function(){$(mbScrollable).mbStopAutoscroll();$(mbScrollable).mbPrevPage();});
			controls.find(".next").bind("click",function(){$(mbScrollable).mbStopAutoscroll();$(mbScrollable).mbNextPage();});
			controls.find(".last").bind("click",function(){$(mbScrollable).mbLastPage();});
			controls.find(".start").bind("click",function(){$(mbScrollable).mbAutoscroll();});
			controls.find(".stop").bind("click",function(){$(mbScrollable).mbStopAutoscroll();mbScrollable.autoScrollActive=false;});
		},

		mbManageControls: function(){
			var mbScrollable=$(this).get(0);
			var controls=$(mbScrollable.options.controls);
			if (mbScrollable.elements.size()<=mbScrollable.options.elementsInPage){
				controls.hide();
			}else{
				controls.show();
			}

			if (mbScrollable.idx==mbScrollable.totalPages){
				controls.find(".last, .next").addClass("disabled");
			}else{
				controls.find(".last, .next").removeClass("disabled");
			}

			if (mbScrollable.idx==1){
				controls.find(".first, .prev").addClass("disabled");
			}else{
				controls.find(".first, .prev").removeClass("disabled");
			}

			if (mbScrollable.autoscroll){
				controls.find(".start").addClass("sel");
				controls.find(".stop").removeClass("sel");
			}else{
				controls.find(".start").removeClass("sel");
				controls.find(".stop").addClass("sel");
			}
			controls.find(".page").removeClass("sel");
			controls.find(".page").eq(mbScrollable.idx-1).addClass("sel");
			controls.find(".idx").html(mbScrollable.idx+" / "+mbScrollable.totalPages);
		},

		goToPage: function(i,noAnim) {
			var mbScrollable= $(this).get(0);
			var anim= noAnim?0:mbScrollable.options.slideTimer;
			if (i>mbScrollable.totalPages) i=mbScrollable.totalPages;
			mbScrollable.scrollTo=-((mbScrollable.singleElDim+mbScrollable.options.elementMargin)*(mbScrollable.options.elementsInPage*(i-1)));
			if(mbScrollable.isVertical){
				if (mbScrollable.scrollTo<-mbScrollable.elementsDim+mbScrollable.options.height)
					mbScrollable.scrollTo=-mbScrollable.elementsDim+mbScrollable.options.height;
				$(mbScrollable.mbscrollableStrip).animate({marginTop:mbScrollable.scrollTo},anim);
			}else{
				if (mbScrollable.scrollTo<-mbScrollable.elementsDim+mbScrollable.options.width)
					mbScrollable.scrollTo=-mbScrollable.elementsDim+mbScrollable.options.width;
				$(mbScrollable.mbscrollableStrip).animate({marginLeft:mbScrollable.scrollTo},anim);
			}
			mbScrollable.idx = Math.floor(i);
			$(mbScrollable).mbManageControls();
			if (!mbScrollable.autoScrollActive)
				$(mbScrollable).mbStopAutoscroll();

			if(mbScrollable.options.changePageCallback)
				mbScrollable.options.changePageCallback(mbScrollable)
		}
	};

	$.fn.mbScrollable=$.mbScrollable.buildMbScrollable;
	$.fn.mbNextPage=$.mbScrollable.mbNextPage;
	$.fn.mbPrevPage=$.mbScrollable.mbPrevPage;
	$.fn.mbFirstPage=$.mbScrollable.mbFirstPage;
	$.fn.mbLastPage=$.mbScrollable.mbLastPage;
	$.fn.mbPageIndex=$.mbScrollable.mbPageIndex;
	$.fn.mbAutoscroll=$.mbScrollable.mbAutoscroll;
	$.fn.mbStopAutoscroll=$.mbScrollable.mbStopAutoscroll;
	$.fn.mbActivateControls=$.mbScrollable.mbActivateControls;
	$.fn.mbManageControls=$.mbScrollable.mbManageControls;
	$.fn.goToPage=$.mbScrollable.goToPage;

})(jQuery);